python/src/main/resources/python/zeppelin_python.py

```diff
@@ -47,46 +35,80 @@ def reset(self):
   def flush(self):
     pass
 
-def handler_stop_signals(sig, frame):
-  sys.exit("Got signal : " + str(sig))
 
+class PythonCompletion:
+  def __init__(self, interpreter, userNameSpace):
+    self.interpreter = interpreter
+    self.userNameSpace = userNameSpace
 
-signal.signal(signal.SIGINT, handler_stop_signals)
+  def getObjectCompletion(self, text_value):
+    completions = [completion for completion in list(self.userNameSpace.keys()) if completion.startswith(text_value)]
+    builtinCompletions = [completion for completion in dir(__builtins__) if completion.startswith(text_value)]
+    return completions + builtinCompletions
 
-host = "127.0.0.1"
-if len(sys.argv) >= 3:
-  host = sys.argv[2]
+  def getMethodCompletion(self, objName, methodName):
+    execResult = locals()
+    try:
+      exec("{} = dir({})".format("objectDefList", objName), _zcUserQueryNameSpace, execResult)
+    except:
+      self.interpreter.logPythonOutput("Fail to run dir on " + objName)
+      self.interpreter.logPythonOutput(traceback.format_exc())
+      return None
+    else:
+      objectDefList = execResult['objectDefList']
+      return [completion for completion in execResult['objectDefList'] if completion.startswith(methodName)]
+
+  def getCompletion(self, text_value):
+    if text_value == None:
+      return None
+
+    dotPos = text_value.find(".")
+    if dotPos == -1:
+      objName = text_value
+      completionList = self.getObjectCompletion(objName)
+    else:
+      objName = text_value[:dotPos]
+      methodName = text_value[dotPos + 1:]
+      completionList = self.getMethodCompletion(objName, methodName)
+
+    if completionList is None or len(completionList) <= 0:
+      self.interpreter.setStatementsFinished("", False)
+    else:
+      result = json.dumps(list(filter(lambda x : not re.match("^__.*", x), list(completionList))))
+      self.interpreter.setStatementsFinished(result, False)
+
+host = sys.argv[1]
+port = int(sys.argv[2])
+
+client = GatewayClient(address=host, port=port)
+gateway = JavaGateway(client, auto_convert = True)
+intp = gateway.entry_point
+# redirect stdout/stderr to java side so that PythonInterpreter can capture the python execution result
+output = Logger()
```

Comment:
there are problems reported with these names too common, conflict with existing variables, output, gateway, client etc, we should name these uniquely if we could

Reply:
It if fine to use them, because they are not in the namespace of user code.

Reply:
I'm not sure - this sets globally right? for example `z` is accessible from user code.

Reply:
User code is in namespace `_zcUserQueryNameSpace` instead of global namespace https://github.com/zjffdu/zeppelin/blob/ZEPPELIN-3375/python/src/main/resources/python/zeppelin_python.py#L91

Reply:
got it